key.go

```diff
@@ -0,0 +1,777 @@
+	}
+}
+
+// Validate ensures that the paramters set inside the Key are internally
+// consistent (e.g. that the key type is appropriate to the curve.
+func (k Key) Validate() error {
+	switch k.KeyType {
+	case KeyTypeEc2:
+		if k.Curve != CurveP256 && k.Curve != CurveP384 && k.Curve != CurveP521 {
+			return fmt.Errorf(
+				"EC2 curve must be P-256, P-384, or P-521; found %q",
+				k.Curve.String(),
+			)
+		}
+	case KeyTypeOkp:
+		if k.Curve != CurveX25519 && k.Curve != CurveX448 &&
+			k.Curve != CurveEd25519 && k.Curve != CurveEd448 {
+
+			return fmt.Errorf(
+				"OKP curve must be X25519, X448, Ed25519, or Ed448; found %q",
+				k.Curve.String(),
+			)
+		}
+	case KeyTypeSymmetric:
+	default:
+		return errors.New(k.KeyType.String())
+	}
+
+	// If Algorithm is set, it must match the specified key parameters.
+	if k.Algorithm != AlgorithmInvalid {
+		expectedAlg, err := k.deriveAlgorithm()
+		if err != nil {
+			return err
+		}
+
+		if k.Algorithm != expectedAlg {
+			return fmt.Errorf(
+				"found algorithm %q (expected %q)",
+				k.Algorithm.String(),
+				expectedAlg.String(),
+			)
+		}
+	}
+
+	return nil
+}
+
+// MarshalCBOR encodes Key into a COSE_Key object.
+func (k Key) MarshalCBOR() ([]byte, error) {
+	var err error
+
+	if k.KeyType == KeyTypeSymmetric {
+		if k.RawKeyParam, err = encMode.Marshal(k.K); err != nil {
+			return nil, err
+		}
+	} else if k.KeyType == KeyTypeEc2 || k.KeyType == KeyTypeOkp {
+		if k.RawKeyParam, err = encMode.Marshal(k.Curve); err != nil {
+			return nil, err
+		}
+	} else {
+		return nil, fmt.Errorf("invalid key type: %q", k.KeyType.String())
+	}
+
+	return encMode.Marshal(k.keyStruct)
+}
+
+// UnmarshalCBOR decodes a COSE_Key object into Key.
+func (k *Key) UnmarshalCBOR(data []byte) error {
+	if err := decMode.Unmarshal(data, &k.keyStruct); err != nil {
+		return err
+	}
+
+	switch k.KeyType {
+	case KeyTypeEc2:
+		if k.RawKeyParam == nil {
+			return errors.New("missing Curve parameter (required for EC2 key type)")
+		}
+
+		if err := decMode.Unmarshal(k.RawKeyParam, &k.Curve); err != nil {
+			return err
+		}
+	case KeyTypeOkp:
+		if k.RawKeyParam == nil {
+			return errors.New("missing Curve parameter (required for OKP key type)")
+		}
+
+		if err := decMode.Unmarshal(k.RawKeyParam, &k.Curve); err != nil {
+			return err
+		}
+	case KeyTypeSymmetric:
+		if k.RawKeyParam == nil {
+			return errors.New("missing K parameter (required for Symmetric key type)")
+		}
+
+		if err := decMode.Unmarshal(k.RawKeyParam, &k.K); err != nil {
+			return err
+		}
+	default:
+		// this should not be reachable as KeyType.UnmarshalCBOR would
+		// result in an error during decMode.Unmarshal() above, if the
+		// value in the data doesn't correspond to one of the above
+		// types.
+		return fmt.Errorf("unexpected key type %q", k.KeyType.String())
+	}
+
+	return k.Validate()
+}
+
+// GetPublic returns a crypto.PublicKey generated using Key's parameters.
+func (k *Key) GetPublic() (crypto.PublicKey, error) {
```

Comment:
It is more idiomatic to omit `Get` from getter methods: https://go.dev/doc/effective_go#Getters